File: src/ruby/bucket_sort.rb
# frozen_string_literal: true

require_relative "./insertion_sort"

# Sort an array using the BucketSort algorithm
class BucketSort
  attr_reader :array_sorted

  def initialize
    @array_sorted = []
    @insertion = InsertionSort.new
  end

  def init(array)
    bucket_sort(array, array.length)
  end

  private

  def bucket_sort(array, bucket_size)
    return nil if array.empty?

    min, max = array.minmax
    bucket_count = ((max - min) / bucket_size).floor + 1
    buckets = Array.new(bucket_count)

    (0..buckets.length - 1).each do |i|
      buckets[i] = []
    end
    (0..array.length - 1).each do |i|
      buckets[((array[i] - min) / bucket_size).floor].push(array[i])
    end
    array.clear
    (0..buckets.length - 1).each do |i|
      @insertion.init(buckets[i])
      buckets[i].each do |value|
        array.push(value)
      end
    end
    @array_sorted = array
  end
end

# test
bu_s = BucketSort.new
bu_s.init([1, 4, 10, 2, 3, 32, 0])
p bu_s.array_sorted

File: src/python/bucket_sort.py
"""Bucket Sort in Python"""


def bucket_sort(vector):
    """
    Implementation of bucket sort.

    Args:
        vector (list): list to be sorted.

    Returns:
        list: Returns the sorted list.

    Example:
    >>> bucket_sort([0.897, 0.565, 0.656, 0.1234, 0.665, 0.3434])
    [0.1234, 0.3434, 0.565, 0.656, 0.665, 0.897]
    """
    if len(vector) == 0:
        return vector

    # Create empty buckets
    bucket_count = len(vector)
    max_val = max(vector)
    min_val = min(vector)

    # Range of each bucket
    bucket_range = (max_val - min_val) / bucket_count
    if bucket_range == 0:
        return vector

    buckets = [[] for _ in range(bucket_count + 1)]

    # Put elements into buckets
    for i in range(len(vector)):
        diff = (vector[i] - min_val) / bucket_range
        index = int(diff)
        buckets[index].append(vector[i])

    # Sort individual buckets and concatenate
    sorted_vector = []
    for i in range(len(buckets)):
        # Using insertion sort for each bucket
        buckets[i] = insertion_sort(buckets[i])
        for j in range(len(buckets[i])):
            sorted_vector.append(buckets[i][j])

    return sorted_vector


def insertion_sort(vector):
    """
    Implementation of insertion sort for use within bucket sort.

    Args:
        vector (list): list to be sorted.

    Returns:
        list: Returns the sorted list.
    """
    for i in range(1, len(vector)):
        key = vector[i]
        aux = i - 1
        while aux >= 0 and vector[aux] > key:
            vector[aux + 1] = vector[aux]
            aux -= 1
        vector[aux + 1] = key
    return vector


if __name__ == "__main__":
    list1 = [0.897, 0.565, 0.656, 0.1234, 0.665, 0.3434]
    print("Unsorted list: ", list1)

    sorted_list = bucket_sort(list1)
    print("Sorted list with bucket sort: ", sorted_list)
